src/test/app/MPToken_test.cpp

```diff
@@ -0,0 +1,1936 @@
+        mptAlice.create(
+            {.ownerCount = 1, .holderCount = 0, .flags = tfMPTCanLock});
+
+        mptAlice.authorize({.account = bob, .holderCount = 1});
+
+        // locks bob's mptoken
+        mptAlice.set({.account = alice, .holder = bob, .flags = tfMPTLock});
+
+        // trying to lock bob's mptoken again will still succeed
+        // but no changes to the objects
+        mptAlice.set({.account = alice, .holder = bob, .flags = tfMPTLock});
+
+        // alice locks the mptissuance
+        mptAlice.set({.account = alice, .flags = tfMPTLock});
+
+        // alice tries to lock up both mptissuance and mptoken again
+        // it will not change the flags and both will remain locked.
+        mptAlice.set({.account = alice, .flags = tfMPTLock});
+        mptAlice.set({.account = alice, .holder = bob, .flags = tfMPTLock});
+
+        // alice unlocks bob's mptoken
+        mptAlice.set({.account = alice, .holder = bob, .flags = tfMPTUnlock});
+
+        // locks up bob's mptoken again
+        mptAlice.set({.account = alice, .holder = bob, .flags = tfMPTLock});
+
+        // alice unlocks mptissuance
+        mptAlice.set({.account = alice, .flags = tfMPTUnlock});
+
+        // alice unlocks bob's mptoken
+        mptAlice.set({.account = alice, .holder = bob, .flags = tfMPTUnlock});
+
+        // alice unlocks mptissuance and bob's mptoken again despite that
+        // they are already unlocked. Make sure this will not change the
+        // flags
+        mptAlice.set({.account = alice, .holder = bob, .flags = tfMPTUnlock});
+        mptAlice.set({.account = alice, .flags = tfMPTUnlock});
+    }
+
+    void
+    testPayment(FeatureBitset features)
+    {
+        testcase("Payment");
+
+        using namespace test::jtx;
+        Account const alice("alice");  // issuer
+        Account const bob("bob");      // holder
+        Account const carol("carol");  // holder
+
+        // preflight validation
+
+        // MPT is disabled
+        {
+            Env env{*this, features - featureMPTokensV1};
+            Account const alice("alice");
+            Account const bob("bob");
+
+            env.fund(XRP(1'000), alice);
+            env.fund(XRP(1'000), bob);
+            STAmount mpt{MPTIssue{makeMptID(1, alice)}, UINT64_C(100)};
+
+            env(pay(alice, bob, mpt), ter(temDISABLED));
+        }
+
+        // MPT is disabled, unsigned request
+        {
+            Env env{*this, features - featureMPTokensV1};
+            Account const alice("alice");  // issuer
+            Account const carol("carol");
+            auto const USD = alice["USD"];
+
+            env.fund(XRP(1'000), alice);
+            env.fund(XRP(1'000), carol);
+            STAmount mpt{MPTIssue{makeMptID(1, alice)}, UINT64_C(100)};
+
+            Json::Value jv;
+            jv[jss::secret] = alice.name();
+            jv[jss::tx_json][jss::Fee] = to_string(env.current()->fees().base);
+            jv[jss::tx_json] = pay(alice, carol, mpt);
```

Comment:
This will overwrite `Fee`.
```suggestion
            jv[jss::tx_json] = pay(alice, carol, mpt);
            jv[jss::tx_json][jss::Fee] = to_string(env.current()->fees().base);
```